design/continuous-delivery.md

```diff
@@ -0,0 +1,576 @@
+
+* Source file or directory (relative to cloud assembly)
+* Packaging format (zip/file)
+* Destinations:
+  * Bucket name
+  * Object key
+  * Publishing Role ARN to assume
+
+For image assets:
+
+* Source directory (relative to cloud assembly): where Dockerfile resides
+* Docker build arguments (optional)
+* Dockerfile name (optional)
+* Destinations:
+  * ECR repository name
+  * Image name
+  * Publishing Role ARN to assume
+
+**Customizability**
+
+Asset consumption and publishing locations should be fully customizable. To that end, the `core.Stack` base class will expose an API that will be used by the asset framework to resolve consumption locations and synthesize `assets.json`. This will allow users to provide their own implementation based on their specific needs.
+
+**Asset Source Staging**
+
+During synthesis, the CDK app will _copy the sources_ of all assets from their original location on disk into the cloud assembly output directory. This allows the cloud assembly to be self-contained and is important for the CI/CD use cases (both internal and external) where the cloud assembly is the only artifact needed after build is complete.
+
+**Asset Providers**
+
+Users should be able to vend custom asset providers to allow customizing how assets are being referenced or prepared.
+
+For example, a company might have an internal system that manages software artifacts. They can internally vend custom implementations for the `lambda.Code` and `ecs.ContainerImage` classes which will allow users to reference these artifacts and synthesize placeholders into the cloud assembly, which will later be resolved during the publishing stage and identified through a user-defined unique identifier.
+
+**Environment-agnostic Stacks**
+
+When a stack is defined, users can specify `env` (account and/or region).
+
+If `account` and/or `region` use the pseudo references `Aws.ACCOUNT_ID` and `Aws.REGION`, respectively, the stack is called "environment-agnostic". Certain features in the CDK, like VPC lookups for example, are not supported for environment-agnostic stacks since the specific account/region is required during synthesis.
+
+The proposal described in [PR#4131](https://github.com/aws/aws-cdk/pull/4131) suggests that environment-agnostics stacks cannot be deployed using the CDK. However, it also proposes that the default behavior for `env` will be to use ("inherit") the CLI configured environment when a stack is deployed through `cdk deploy`.
+
+This means that the only way to produce environment-agnostic templates will be to explicitly indicate it when a stack is defined.
+
+Since the specific account and region are required when resolving asset consumption and publishing locations, the current plan is for the default asset store implementation to **fail if assets are used from environment-agnostic stacks**. Again, bear in mind that the current behavior (where the default is environment-agnostic stacks) is going to be changed.
+
+## Bootstrapping
```

Comment:
Feedback from @bundyfx:

- Ability to add/remove trusted deployment accounts
- Ability to control the policy of the CloudFormation deployment role
- Ability to provision a large set of environments with multiple types of policies (e.g. dev/test/prod).

Reply:
@bundyfx also described a "bootstrapping pipeline" that resides in the master account of the AWS Organization, discovers new accounts, assumes the administrative organization role in those accounts and bootstraps them.